### ISSUE TITLE
Plan to put new ocds-cove repository live

### ISSUE THREAD

**Issue by odscjames** [open-contracting/deploy#97]
Goal: Putting https://github.com/open-contracting/cove-ocds live instead of https://github.com/OpenDataServices/cove

We were looking for a way to do this that:
  * Allowed for lots of testing before going live by many staff
  * Any downtime is gone, minimised or at least controlled well
  * Has good rollback options if needed

We currently have: "live.cove.opencontracting.uk0.bigv.io" server, serving Cove, Docs server proxies to this.

The current state of the salt scripts don't make it easy to have more than one cove on the same server, but we can easily have a separate server with a different cove (as we already do for oc4ids).

We would like to suggest:

* We spin up a new server - install https://github.com/open-contracting/cove-ocds with a new pillar variable file. Server is low spec to keep costs down.
* Allow testing by staff for final approval before switchover.
* Shut down new server, scale up CPUs and RAM to same as old server, restart.
* Go to Docs server, switch which cove backend we use to new server.
* Shut down old server, scale down CPUs and RAM to keep costs down. Leave for a month. If  we need to rollback in a hurry it's ready. Otherwise, delete.

Does this basically sound good?

One drawback with this is that it doesn't preserve data - if someone uploads a file 5 mins before switchover then sends the link to someone who looks 10 minutes later, the file will be gone.  If we need a plan that avoids this, then something like: (I was unsure what level of service people expect with this, and how much notification we should provide if any?)

* We spin up a new server - install https://github.com/open-contracting/cove-ocds with a new pillar variable file. Server is low spec to keep costs down.
* Allow testing by staff for final approval before switchover.
* Shut down new server, scale up CPUs and RAM to same as old server, restart.
* Go to Docs server, switch which cove backend we use to point to a dummy placeholder page saying "It's down"
* Copy db.sqlite3 and media folder from old server to new server. It's currently about ~400MB so minutes.
* Go to Docs server, switch which cove backend we use to point to new server
* Shut down old server, scale down CPUs and RAM to keep costs down. Leave for a month. If  we need to rollback in a hurry it's ready. Otherwise, delete.

**Comment by jpmckinney** [comment]
Would being able to have two CoVEs on the same server simplify this? It wouldn't take long to restore functionality for multiple versions of the same app on one server.

**Comment by odscjames** [comment]
I don't think so actually; or at least not by an amount we'd actually notice. Starting a new machine or using SCP is routine, and that's balanced against the fact that on the same machine we'd have the added complexity of rewriting salt. I think it's best to keep the salt stable and just go with a new machine.

How much notice / down time messages do you think we need for this service?

**Comment by jpmckinney** [comment]
Okay, I think ODS has access to Bytemark to create the server.

In the first option there's no downtime, and in the second option there's only minutes (less than an hour anyway) with a clear maintenance message (which can say the time by which the service is expected to return, e.g. "We are performing maintenance. The service is expected to return by 2020-01-DD HH:MM"). Since it's not mission-critical for it to be always online, I think it's fine to do either option without notice (I'm also not sure to whom we'd send notice?).

**Comment by jpmckinney** [comment]
I forgot to note: my preference is for the second option (two additional steps).

Since we'll be creating a new server, can we take this opportunity to document the steps? https://github.com/open-contracting/deploy/issues/16 Alternately, we can do it together in a screen share (like for my first Salt deploy), and use that experience to write initial documentation.

**Comment by rhiaro** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Refactor OCDS-specific code into a new application](https://trello.com/c/6SyOfmT6/49-refactor-ocds-specific-code-into-a-new-application)

**Comment by odscjames** [comment]
Server created. Others will pass on details soon.

As you were out today, I just did this and wrote up - will comment in #16 . It's a pretty easy process - that's the point of salt :-)

**Comment by odscjames** [comment]
Changeover done

Downtime:
* A 4 min batch when I broke a "if testing"  clause when trying to get a downtime page working (Ironic)
* There was a problem with AliasMatch when something else was interfering with it. After a bit, I decided it would take longer to try and work out the problem that it was worth. I just did it, with 2 mins downtime
* A Django setting had not been copied across and so "Convert to Spreadsheet" was broken. This was soon fixed.

Now leaving for a bit to make sure no more reports of problems come in.

New server set to same high specs as old server, old server turned right down to cheap specs.
